hollow/src/main/java/com/netflix/hollow/tools/history/keyindex/HollowOrdinalMapper.java

```diff
@@ -0,0 +1,196 @@
+
+        // Linear probing
+        while (ordinalMappings[index] != ORDINAL_NONE) {
+            if(recordsAreEqual(typeState, ordinal, index)) {
+                this.assignedOrdinalToIndex.put(assignedOrdinal, index);
+                return ORDINAL_NONE;
+            }
+            index = (index + 1) % ordinalMappings.length;
+        }
+
+        ordinalMappings[index] = assignedOrdinal;
+        originalHash[index] = hashedRecord;
+        size++;
+
+        storeFields(typeState, ordinal, index);
+
+        this.assignedOrdinalToIndex.put(assignedOrdinal, index);
+        return index;
+    }
+
+    private void storeFields(HollowObjectTypeReadState typeState, int ordinal, int index) {
+        if(!indexFieldObjectMapping.containsKey(index))
+            indexFieldObjectMapping.put(index, new Object[primaryKey.numFields()]);
+
+        for(int i=0;i<primaryKey.numFields();i++) {
+            if(!keyFieldIsIndexed[i])
+                continue;
+
+            Object objectToStore = readValueInState(typeState, ordinal, i);
+            indexFieldObjectMapping.get(index)[i] = memoizedPool.intern(objectToStore);
+        }
+    }
+
+    private void expandAndRehashTable() {
+        Integer[] newTable = new Integer[ordinalMappings.length*2];
+        Arrays.fill(newTable, ORDINAL_NONE);
+
+        Integer[] newOriginalHash = new Integer[originalHash.length*2];
+        HashMap<Integer, Object[]> newIndexFieldObjectMapping = new HashMap<>();
+
+        for(int i=0;i<ordinalMappings.length;i++) {
+            if(ordinalMappings[i]==ORDINAL_NONE)
+                continue;
+            int newIndex = rehashExistingRecord(newTable, originalHash[i], ordinalMappings[i]);
+            newOriginalHash[newIndex] = originalHash[i];
+
+            Object[] fieldObjects = indexFieldObjectMapping.get(i);
+            newIndexFieldObjectMapping.put(newIndex, fieldObjects);
+        }
+
+        this.ordinalMappings = newTable;
+        this.originalHash = newOriginalHash;
+        this.indexFieldObjectMapping = newIndexFieldObjectMapping;
+    }
+
+    private int rehashExistingRecord(Integer[] newTable, int originalHash, int assignedOrdinal) {
+        int newIndex = indexFromHash(originalHash, newTable.length);
+        while (newTable[newIndex]!=ORDINAL_NONE)
+            newIndex = (newIndex + 1) % newTable.length;
+
+        assignedOrdinalToIndex.put(assignedOrdinal, newIndex);
+        newTable[newIndex] = assignedOrdinal;
+        return newIndex;
+    }
+
+    public Object getFieldObject(int keyOrdinal, int fieldIndex) {
+        int index = assignedOrdinalToIndex.get(keyOrdinal);
+        return indexFieldObjectMapping.get(index)[fieldIndex];
+    }
+
+    private int hashKeyRecord(HollowObjectTypeReadState typeState, int ordinal) {
+        int hashCode = 0;
+        for (int i = 0; i < primaryKey.numFields(); i++) {
+            int fieldHashCode = HollowReadFieldUtils.fieldHashCode(typeState, ordinal, i);
```

Comment:
I think there is a regression here for when the field at index i is of REFERENCE type. In that case, the hash is computed based on the value of the ordinal into the referenced type, and over time that ordinal could be reused for a different value, resulting in hashes not being consistently computed. E.g. data model-

```
Movie @PrimaryKey(id) {
    Video id;
    String name;
}

Video {
    int value;
}
```

I think this explains the corruption type behavior I was seeing where the diff counts didn't match with baseline. I've attempted a patch in [this commit](https://github.com/Netflix/hollow/commit/43db4ba9f20b3135fe6f151e70422b715629a669) but haven't tested it yet

Reply:
I see, I fixed this in other places but missed it here. Good catch! I think I already have a function for this, will LYK

Reply:
Interesting, I discussed this exact issue with Drew and he suggested using ordinals (as opposed to values) because the records are immutable. However, I'm not sure if he considered ordinals being reused, and your theory makes sense to me.

I already have a function elsewhere that retrieves the referenced value, I'll implement that to make it a bit cleaner.

Reply:
Fixed in 51cc570d